### ISSUE TITLE
Memory leak: UGen Ctors get called over and over again when they throw exceptions (e.g. alloc failed)

### ISSUE THREAD

**Issue by elgiano** [supercollider/supercollider#5634]
## Environment

* SuperCollider version: 3.12
* Operating system: Linux
* Other details (Qt version, audio driver, etc.): Jack

## Steps to reproduce
I encounter this using PartConv, but I suspect it's something more general.
```supercollider
// allocate a big IR buffer
b = Buffer.alloc(s, PartConv.calcBufSize(2048, (numFrames: 44100 * 300)).asInteger);
// this is too much for my 8GB ram, trying to play PartConv causes alloc failed
x = { PartConv.ar(Silent.ar, 2048, b) }.play;
// if x runs for a little while, it eventually prints "server disconnected shared memory interface"
x.free;
// then if I try to allocate anything else, I still get "alloc failed"
{DelayN.ar(Silent.ar, 1)}.play
```

on supernova:

```supercollider
{DelayN.ar(Silent.ar, 100)}.play
// -> Failed to allocate memory for  ugen.
// Server 'localhost' exited with exit code 0.
// server 'localhost' disconnected shared memory interface
```

## Expected vs. actual behavior
I would expect `x.free` to cleanup as if the "too big" memory allocation never happened, so that I can still allocate a smaller amount of memory.
What I observe is that on failed allocation, the UGen constructor runs repeteadly (once for every time the alloc failed error is printed), and the destructor never runs (I observed this by adding a few Print statements in UGens constructors and destructors). Destructors are not called even when the synth or the whole Node Tree is freed.
This is by the way how error messages are "flooding" (relating to #3563): the error occurs again and again at every processing block.

I observed this looping behavior also with DelayN, but I can't get to the "disconnected shared memory interface" point with DelayN.

On supernova the server crashes when I try to allocate too much memory.... I assume it might be linked but it looks also like a different issue to me.

**Comment by elgiano** [comment]
What I've figured out so far:

- When a new Synth is created, scsynth calls Graph_New, which sets the graph's mCalcFunc to Graph_FirstCalc
- Graph_FirstCalc calls all the UGens constructors, and then sets the graph's mCalcFunc to Graph_Calc
- If any of the Ctors fails by throwing an exception (such as the alloc failed error), Graph_FirstCalc exits without changing the graph's mCalcFunc, so FirstCalc gets called again at the next processing block, calling again all constructors

This means that if a UGen's Ctor throws an exception, all Ctors for that UGen will be called again and again at each block. Since the Graph gets stuck at FirstCalc, Dtors are never called, so if any of those Ctors does any RTAlloc, memory gets allocated over and over again, it never gets freed and references to it get lost. It looks like a quite scary memory leak, isn't it?

Now, I put a try/catch block in PartConv locally, so that I can keep working, but doesn't this issue deserve a more general treatment, if possible? I have no concrete idea about how to address it, and it looks like a quite complex issue since I imagine we don't want to put a try/catch block in FirstCalc... I think Ctors that call functions that can throw exceptions (like RTAlloc) should be responsible for catching them and cleaning up.

In case it could be useful to anyone else, my dirty fix for PartConv looks like this:

```diff 
diff --git a/server/plugins/PartitionedConvolution.cpp b/server/plugins/PartitionedConvolution.cpp
index 1bf63d8ed..060722172 100644
--- a/server/plugins/PartitionedConvolution.cpp
+++ b/server/plugins/PartitionedConvolution.cpp
@@ -179,8 +179,16 @@ void PartConv_Ctor(PartConv* unit) {
         // unit->m_partitions,unit->m_nover2, unit->m_blocksize, unit->m_spareblocks, unit->m_numamort,
         // unit->m_lastamort);
 
-        unit->m_fd_accumulate = (float*)RTAlloc(unit->mWorld, unit->m_fullsize * sizeof(float));
-        memset(unit->m_fd_accumulate, 0, unit->m_fullsize * sizeof(float));
+        try{
+          unit->m_fd_accumulate = (float*)RTAlloc(unit->mWorld, unit->m_fullsize * sizeof(float));
+          memset(unit->m_fd_accumulate, 0, unit->m_fullsize * sizeof(float));
+        } catch(std::exception const& e) {
+          printf("PartConv: exception in real time: %s\n", e.what());
+          SETCALC(*ClearUnitOutputs);
+          unit->mDone = true;
+          return;
+        }
+
         unit->m_fd_accum_pos = 0;
 
         SETCALC(PartConv_next);
```

**Comment by jamshark70** [comment]
UGen constructors running repeatedly is definitely bad. Setting the function to clear outputs is better. Whether that's done in a try/catch or by some other mechanism, I'm not qualified to say. (Though, if the RTAlloc exception doesn't crash the server, then doesn't this mean there's a try block somewhere up the call stack? I wonder why that error handler isn't updating the calculation function.)

Note though that 1/ Buffer.alloc doesn't touch the RT memory pool and b/ the fact that your system has 8GB RAM has nothing to do with PartConv's memory allocation (which is governed by server options memSize -- I think this issue should be reproducible with a much smaller kernel, if memSize is restricted to an unrealistically small amount).

**Comment by elgiano** [comment]
The exception is caught by the "AudioDriver", in my case JackDriver, after the processing block function. This means that all UGen calculations are stopped, and we skip to the next block (where the same error will happen again). By the way, this would happen also for UGens failing at RTAlloc during their `_next` function.

I "like" the idea proposed by @nhthn in #782. It looks like a macro is the way to go to fix "older" UGens. For newer UGens we could have a function in SC_Plugin.hpp... I start to see how this requires an RFC, I'll make one later.

**Comment by elgiano** [comment]
Fixed by  #5713
